src/testing-policies.md

```diff
@@ -1,61 +1,168 @@
+on top of a Kubernetes cluster, we have to provide the resource type
+that it would receive from the API server: `AdmissionReview` API
+object types.
+
+Let's create two different requests, one that is valid, while the
+other is invalid.
+
+This is a valid `AdmissionReview.request` object according to the
+policy that we want to test:
+
+```json
+{
+  "kind": {
+    "kind": "Pod",
+    "version": "v1"
+  },
+  "object": {
+    "metadata": {
+      "name": "valid-pod"
+    },
+    "spec": {
+      "containers": [
+        {
+          "image": "nginx",
+          "name": "valid-container"
+        }
+      ]
+    }
+  },
+  "operation": "CREATE",
+  "requestKind": {"version": "v1", "kind": "Pod"},
+  "userInfo": {
+    "username": "some-user",
+    "uid": "some-uid",
+    "groups": ["system:authenticated", "group-a", "group-b"]
+  }
+}
+```
 
-This is a slightly redacted snippet taken from the `Makefile` of the project:
+This is an invalid `AdmissionReview.request` object according to the
+policy that we want to test:
+
+```json
+{
+  "kind": {
+    "kind": "Pod",
+    "version": "v1"
+  },
+  "object": {
+    "metadata": {
+      "name": "invalid-pod"
+    },
+    "spec": {
+      "containers": [
+        {
+          "image": "nginx",
+          "name": "invalid-container",
+          "securityContext": {
+            "privileged": true
+          }
+        }
+      ]
+    }
+  },
+  "operation": "CREATE",
+  "requestKind": {"version": "v1", "kind": "Pod"},
+  "userInfo": {
+    "username": "some-user",
+    "uid": "some-uid",
+    "groups": ["system:authenticated", "group-a", "group-b"]
+  }
+}
+```
 
-```shell
-echo Accepting policy
-hyperfine --warmup 10 "cat test_data/req_pod_with_equal_toleration.json | wasmtime run --env TAINT_KEY="dedicated-key" --env TAINT_VALUE="tenantA" --env ALLOWED_GROUPS="system:authenticated" target/wasm32-wasi/release/pod-toleration-policy.wasm"
+### Test the policy
 
-echo Rejecting policy
-hyperfine --warmup 10 "cat test_data/req_pod_with_equal_toleration.json | wasmtime run --env TAINT_KEY="dedicated-key" --env TAINT_VALUE="tenantA" --env ALLOWED_GROUPS="tenantA-users" target/wasm32-wasi/release/pod-toleration-policy.wasm"
+Now we can use `chimera-policy-testdrive` to test both requests:
 
-echo Operation not relevant
-hyperfine --warmup 10 "cat test_data/req_delete.json | wasmtime run --env TAINT_KEY="dedicated-key" --env TAINT_VALUE="tenantA" --env ALLOWED_GROUPS="tenantA-users" target/wasm32-wasi/release/pod-toleration-policy.wasm"
+```console
+$ chimera-policy-testdrive --policy module.wasm --request-file valid-admission-review.json
+ValidationResponse { accepted: true, message: Some(""), code: None }
 ```
+
+Whereas if we execute the invalid request, we will get:
+
+```console
+$ chimera-policy-testdrive --policy module.wasm --request-file invalid-admission-review.json
+ValidationResponse { accepted: false, message: Some("User cannot schedule privileged containers"), code: None }
+```
+
```

Comment:
I would also add an example where we pass some settings to the policy. Like the users that are trusted.